File: android/app/src/main/java/com/betterlise/app/domain/AgendaLayout.kt
package com.betterlise.app.domain

import com.betterlise.app.data.api.CalendarEvent
import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.TemporalAdjusters

val PARIS: ZoneId = ZoneId.of("Europe/Paris")

/** Column placement of a timed event inside a day timeline. */
data class PlacedEvent(val event: CalendarEvent, val column: Int, val columnCount: Int)

object AgendaLayout {
    /**
     * Groups overlapping events into clusters and assigns each event the first free column
     * of its cluster. Every event in a cluster shares the cluster's column count.
     */
    fun place(events: List<CalendarEvent>): List<PlacedEvent> {
        val sorted = events
            .filterNot { it.isAllDay }
            .sortedWith(compareBy<CalendarEvent> { it.startDate }.thenBy { it.endDate })

        val placed = mutableListOf<PlacedEvent>()
        val cluster = mutableListOf<Pair<CalendarEvent, Int>>()
        val columnEnds = mutableListOf<Instant>()
        var clusterEnd = Instant.MIN

        fun flush() {
            val count = maxOf(columnEnds.size, 1)
            cluster.forEach { (event, column) -> placed += PlacedEvent(event, column, count) }
            cluster.clear()
            columnEnds.clear()
        }

        for (event in sorted) {
            if (cluster.isNotEmpty() && event.startDate >= clusterEnd) flush()
            val free = columnEnds.indexOfFirst { it <= event.startDate }
            if (free >= 0) {
                columnEnds[free] = event.endDate
                cluster += event to free
            } else {
                columnEnds += event.endDate
                cluster += event to columnEnds.lastIndex
            }
            if (event.endDate > clusterEnd) clusterEnd = event.endDate
        }
        flush()
        return placed
    }

    /**
     * Every school day (Monday–Friday) of the week of [reference] and the surrounding weeks.
     * On weekends the reference week is the upcoming one.
     */
    fun schoolDays(reference: LocalDate, weeksBefore: Int, weeksAfter: Int): List<LocalDate> {
        val anchor = if (reference.dayOfWeek.value >= 6) reference.with(TemporalAdjusters.next(DayOfWeek.MONDAY)) else reference
        val monday = anchor.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))
        return (-weeksBefore..weeksAfter).flatMap { week ->
            (0L until 5L).map { day -> monday.plusWeeks(week.toLong()).plusDays(day) }
        }
    }

    /** Index of [today] in [days], or of the next school day when today is a weekend. */
    fun initialIndex(days: List<LocalDate>, today: LocalDate): Int =
        days.indexOfFirst { !it.isBefore(today) }.takeIf { it >= 0 } ?: (days.size - 1).coerceAtLeast(0)

    /** Events keyed by their Paris calendar day, sorted by start time. */
    fun groupByDay(events: List<CalendarEvent>): Map<LocalDate, List<CalendarEvent>> =
        events.groupBy { it.startDate.atZone(PARIS).toLocalDate() }.mapValues { (_, dayEvents) -> dayEvents.sortedBy { it.startDate } }

    /** An event is past once it has ended; the week view dims those. */
    fun isPast(event: CalendarEvent, now: Instant): Boolean = event.endDate <= now

    /** Whether the hour label of [hour] would collide with the current time shown in the gutter. */
    fun isHourLabelNearNow(hour: Int, now: Instant): Boolean = now.atZone(PARIS).let { time ->
        kotlin.math.abs(time.hour * 60 + time.minute - hour * 60) < 15
    }

    /** Paris time in the style of the hour gutter: "8h", "13h30". */
    fun shortTime(instant: Instant): String = instant.atZone(PARIS).let { time ->
        "${time.hour}h" + if (time.minute == 0) "" else "%02d".format(time.minute)
    }

    fun eventsOn(day: LocalDate, events: List<CalendarEvent>): List<CalendarEvent> =
        events.filter { it.startDate.atZone(PARIS).toLocalDate() == day }
}

File: ios/BetterLise/Features/Agenda/AgendaLayout.swift
import Foundation

/// Column placement of a timed event inside a day timeline.
struct PlacedEvent: Identifiable, Equatable {
    let event: CalendarEvent
    let column: Int
    let columnCount: Int

    var id: String { event.id }
}

enum AgendaLayout {
    /// Groups overlapping events into clusters and assigns each event the first free
    /// column of its cluster. Every event in a cluster shares the cluster's column count.
    static func place(_ events: [CalendarEvent]) -> [PlacedEvent] {
        let sorted = events
            .filter { !$0.isAllDay }
            .sorted { ($0.startDate, $0.endDate) < ($1.startDate, $1.endDate) }

        var placed: [PlacedEvent] = []
        var cluster: [(event: CalendarEvent, column: Int)] = []
        var columnEnds: [Date] = []
        var clusterEnd = Date.distantPast

        func flushCluster() {
            let count = max(columnEnds.count, 1)
            placed += cluster.map { PlacedEvent(event: $0.event, column: $0.column, columnCount: count) }
            cluster = []
            columnEnds = []
        }

        for event in sorted {
            if event.startDate >= clusterEnd, !cluster.isEmpty {
                flushCluster()
            }
            if let free = columnEnds.firstIndex(where: { $0 <= event.startDate }) {
                columnEnds[free] = event.endDate
                cluster.append((event, free))
            } else {
                columnEnds.append(event.endDate)
                cluster.append((event, columnEnds.count - 1))
            }
            clusterEnd = max(clusterEnd, event.endDate)
        }
        flushCluster()
        return placed
    }

    /// Monday–Friday of the week containing `reference`, shifted by `weekOffset` weeks.
    static func weekDays(containing reference: Date, weekOffset: Int, calendar: Calendar = .paris) -> [Date] {
        guard let shifted = calendar.date(byAdding: .weekOfYear, value: weekOffset, to: reference),
              let monday = calendar.dateInterval(of: .weekOfYear, for: shifted)?.start else {
            return []
        }
        return (0..<5).compactMap { calendar.date(byAdding: .day, value: $0, to: monday) }
    }

    static func events(on day: Date, from events: [CalendarEvent], calendar: Calendar = .paris) -> [CalendarEvent] {
        events.filter { calendar.isDate($0.startDate, inSameDayAs: day) }
    }
}
